apis/management.cattle.io/v3/rke_types.go

```diff
@@ -337,14 +337,146 @@ type PortCheck struct {
 type CloudProvider struct {
 	// Name of the Cloud Provider
 	Name string `yaml:"name" json:"name,omitempty"`
-	// Configuration Options of Cloud Provider
-	CloudConfig map[string]string `yaml:"cloud_config" json:"cloudConfig,omitempty"`
-	// AWSCloudProvicer
-	AWSCloudProvider *AWSCloudProvider `yaml:",omitempty" json:"awsCloudProvider,omitempty"`
-	// AzureCloudProvicer
-	AzureCloudProvider *AzureCloudProvider `yaml:",omitempty" json:"azureCloudProvider,omitempty"`
+	// AWSCloudProvider
+	AWSCloudProvider *AWSCloudProvider `yaml:"awsCloudProvider,omitempty" json:"awsCloudProvider,omitempty"`
+	// AzureCloudProvider
+	AzureCloudProvider *AzureCloudProvider `yaml:"azureCloudProvider,omitempty" json:"azureCloudProvider,omitempty"`
+	// OpenstackCloudProvider
+	OpenstackCloudProvider *OpenstackCloudProvider `yaml:"openstackCloudProvider,omitempty" json:"openstackCloudProvider,omitempty"`
+	// VsphereCloudProvider
+	VsphereCloudProvider *VsphereCloudProvider `yaml:"vsphereCloudProvider,omitempty" json:"vsphereCloudProvider,omitempty"`
+	// CustomCloudProvider
+	CustomCloudProvider string `yaml:"customCloudProvider,omitempty" json:"customCloudProvider,omitempty"`
```

Comment:
what is custom cloud provider? does it come with 0 options?

Reply:
Custom cloud provider is just multi line string, where the user can put a custom config file

Reply:
Can we add it to the comments?

Reply:
sure